projects/ad405x_iio/app/ad405x_iio.c

```diff
@@ -0,0 +1,1576 @@
+			       ad405x_burst_sample_rates_str[15]);
+#endif
+
+	case ADC_FILTER_LENGTH:
+		return sprintf(buf,
+			       "%s %s %s %s %s %s %s %s"
+#if !defined(DEV_AD4050)
+			       " %s %s %s %s",
+#else
+			       ,
+#endif
+			       ad405x_avg_filter_str[0],
+			       ad405x_avg_filter_str[1],
+			       ad405x_avg_filter_str[2],
+			       ad405x_avg_filter_str[3],
+			       ad405x_avg_filter_str[4],
+			       ad405x_avg_filter_str[5],
+			       ad405x_avg_filter_str[6],
+			       ad405x_avg_filter_str[7]
+#if !defined(DEV_AD4050)
+			       ,ad405x_avg_filter_str[8],
+			       ad405x_avg_filter_str[9],
+			       ad405x_avg_filter_str[10],
+			       ad405x_avg_filter_str[11]
+#endif
+			      );
+#endif
+
+	default:
+		break;
+	}
+
+	return len;
+}
+
+/*!
+ * @brief	Attribute available setter function for ad405x attributes
+ * @param	device[in, out]- Pointer to IIO device instance
+ * @param	buf[in]- IIO input data buffer
+ * @param	len[in]- Number of input bytes
+ * @param	channel[in] - input channel
+ * @param	priv[in] - Attribute private ID
+ * @return	len in case of success, negative error code otherwise
+ */
+static int iio_ad405x_attr_available_set(void *device,
+		char *buf,
+		uint32_t len,
+		const struct iio_ch_info *channel,
+		intptr_t priv)
+{
+	return len;
+}
+
+/*!
+ * @brief	Start a data capture in continuous/burst mode
+ * @return	0 in case of success, negative error code otherwise
+ */
+static int32_t ad405x_start_data_capture(void)
+{
+	int ret;
+
+#if (ADC_CAPTURE_MODE == SAMPLE_MODE)
+	ret = ad405x_set_adc_mode(p_ad405x_dev);
+#else
+	ret = ad405x_set_burst_averaging_mode(p_ad405x_dev);
+#endif
+	if (ret) {
+		return ret;
+	}
+
+#if (INTERFACE_MODE == SPI_INTERRUPT)
+	ret = no_os_pwm_enable(pwm_desc);
+	if (ret) {
+		return ret;
+	}
+
+#if (APP_CAPTURE_MODE == CONTINUOUS_DATA_CAPTURE)
+	ret = iio_trig_enable(ad405x_hw_trig_desc);
+	if (ret) {
+		return ret;
+	}
+#else
+	ret = no_os_irq_enable(trigger_irq_desc, TRIGGER_INT_ID);
+	if (ret) {
+		return ret;
+	}
+#endif
+#endif
+
+	return 0;
+}
+
+/*!
+ * @brief	Stop a data capture from continuous/burst mode
+ * @return	0 in case of success, negative error code otherwise
+ */
+int32_t ad405x_stop_data_capture(void)
+{
+	int ret;
+
+#if (INTERFACE_MODE == SPI_INTERRUPT)
+	ret = no_os_pwm_disable(pwm_desc);
+	if (ret) {
+		return ret;
+	}
+#endif
+
+#if (APP_CAPTURE_MODE == CONTINUOUS_DATA_CAPTURE)
+	ret = iio_trig_disable(ad405x_hw_trig_desc);
+	if (ret) {
+		return ret;
+	}
+#else
+	ret = no_os_irq_disable(trigger_irq_desc, TRIGGER_INT_ID);
+	if (ret) {
+		return ret;
+	}
+#endif
+
+#if (INTERFACE_MODE == SPI_DMA)
+	/* Abort DMA and Timers and configure CS and CNV as GPIOs */
+	stm32_timer_stop();
+	stm32_abort_dma_transfer();
+	stm32_cs_output_gpio_config(true);
+	stm32_cnv_output_gpio_config(true);
+
+	spi_init_param = ad405x_init_params.spi_init->extra;
+	spi_init_param->dma_init = NULL;
+	ad405x_init_params.spi_init->max_speed_hz = MAX_SPI_SCLK;
```

Comment:
do we need to change spi frequency again?

Reply:
In the config mode, the max SPI SCLK allowed is less than 45MHz, so we do have to change it. For the sample_mode, and bust_averaging modes, the supported SPI clock freq is higher than 45MHz so we use that to minimize the transaction time